tests/unit/onyxClearNativeStorageTest.js

```diff
@@ -0,0 +1,162 @@
+
+    /** @type OnyxCache */
+    let cache;
+
+    // Always use a "fresh" cache instance
+    beforeEach(() => {
+        onyxValue = null;
+        cache = require('../../lib/OnyxCache').default;
+        Onyx.init({
+            keys: ONYX_KEYS,
+            registerStorageEventListener: () => {},
+            initialKeyStates: {
+                [ONYX_KEYS.DEFAULT_KEY]: DEFAULT_VALUE,
+            },
+        });
+        connectionID = Onyx.connect({
+            key: ONYX_KEYS.DEFAULT_KEY,
+            initWithStoredValues: false,
+            callback: val => onyxValue = val,
+        });
+        return waitForPromisesToResolve();
+    });
+
+    afterEach(() => {
+        Onyx.disconnect(connectionID);
+        Storage.clear.mockReset();
+        return Onyx.clear()
+            .then(() => {
+                storageCallResolveList = [];
+                storageCallQueue = [];
+            });
+    });
+
+    it('should persist the value of Onyx.merge when called between the cache and storage clearing', () => {
+        expect.assertions(5);
+
+        // GIVEN that Onyx is completely clear
+        // WHEN Onyx.clear() is called
+        Storage.clear = jest.fn(() => {
+            // WHEN merge is called between the cache and storage clearing, on a key with a default key state
+            Onyx.merge(ONYX_KEYS.DEFAULT_KEY, MERGED_VALUE);
+            AsyncStorageMock.clear();
+            return waitForPromisesToResolve();
+        });
+        Onyx.clear();
+        return waitForPromisesToResolve()
+            .then(() => {
+                // THEN the storage finishes clearing before Storage.setItem finishes for the merged value
+                expect(getStorageCallResolveOrder('clear')).toBe(1);
+                expect(getStorageCallResolveOrder('setItem')).toBe(2);
+
+                // THEN the value in Onyx, the cache, and the storage is the merged value
+                expect(onyxValue).toBe(MERGED_VALUE);
+                const cachedValue = cache.getValue(ONYX_KEYS.DEFAULT_KEY);
+                expect(cachedValue).toBe(MERGED_VALUE);
+                const storedValue = Storage.getItem(ONYX_KEYS.DEFAULT_KEY);
+                return expect(storedValue).resolves.toBe(MERGED_VALUE);
+            });
+    });
+
+    it('should replace the value of Onyx.set with the default key state in the cache', () => {
+        expect.assertions(5);
+
+        // GIVEN that Onyx is completely clear
+        // WHEN set then clear is called on a key with a default key state
+        Onyx.set(ONYX_KEYS.DEFAULT_KEY, SET_VALUE);
+        Storage.clear = jest.fn(() => AsyncStorageMock.clear());
+        Onyx.clear();
+        return waitForPromisesToResolve()
+            .then(() => {
+                // THEN the Storage.setItem finishes for the set value before the storage finishes clearing
+                expect(getStorageCallResolveOrder('setItem')).toBe(1);
+                expect(getStorageCallResolveOrder('clear')).toBe(2);
```

Comment:
Confused about why the order in which these resolve is relevant. Shouldn't it always be the same? What would change this? Are we trying to prevent a specific situation by adding this logic?